src/main/kotlin/org/jitsi/jicofo/conference/source/Source.kt

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright @ 2021 - present 8x8, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jitsi.jicofo.conference.source
+
+import org.jitsi.utils.MediaType
+import org.jitsi.xmpp.extensions.colibri.SourcePacketExtension
+import org.jitsi.xmpp.extensions.jingle.ParameterPacketExtension
+import org.jitsi.xmpp.extensions.jitsimeet.SSRCInfoPacketExtension
+import org.jxmpp.jid.Jid
+
+/**
+ * The description of a single source (i.e. an SSRC and its properties from "a=ssrc" lines in SDP).
+ */
+data class Source(
+    val ssrc: Long,
+    /** Audio or Video **/
+    val mediaType: MediaType,
+    /** Optional msid */
+    val msid: String? = null,
+    /** Optional cname */
+    val cname: String? = null,
+    /**
+     * An internal flag indicating that this source was locally generated by jicofo and isn't/shouldn't be signalled
+     * to endpoints.
+     */
+    val injected: Boolean = false,
+) {
+    /** Create a [Source] from an XML extension. */
+    constructor(mediaType: MediaType, sourcePacketExtension: SourcePacketExtension) : this(
+        sourcePacketExtension.ssrc,
+        mediaType,
+        sourcePacketExtension.getChildExtensionsOfType(ParameterPacketExtension::class.java)
+            .filter { it.name == "msid" }.map { it.value }.firstOrNull(),
+        sourcePacketExtension.getChildExtensionsOfType(ParameterPacketExtension::class.java)
+            .filter { it.name == "cname" }.map { it.value }.firstOrNull(),
+        sourcePacketExtension.isInjected
+    )
+
+    /** Serializes this [SsrcGroup] to XML */
+    fun toPacketExtension(
+        /** An optional JID for the owner of this source to encode in the XML extension. */
+        owner: Jid? = null
+    ) = SourcePacketExtension().apply {
+        ssrc = this@Source.ssrc
+        if (owner != null) {
+            SSRCInfoPacketExtension().also {
+                it.owner = owner
+                addChildExtension(it)
```

Comment:
I find calling `addChildExtension(it)` in the `also` a bit confusing - maybe it'd be clearer to call `addChildExtension` on the `SSRCInfoPacketExtension` constructor?

Reply:
Agreed and fixed.